lib: Add nexthop_cmp

sworleys opened the pull request:
Add function to allow us to have a sorted order
of nexthops.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

mjstapp commented on lib/nexthop.c:
```diff
@@ -36,6 +36,63 @@
+	int ret;
+	uint32_t n1, n2;
+
+	if (next1->vrf_id < next2->vrf_id)
```
Is this restricted to {-1, 0, 1} ? if not, can we just subtract the vrf_ids?

sworleys replied:
You mean this?
```
if (next1->vrf_id - next2->vrf_id)
     return (next1->vrf_id - next2->vrf_id)
```
If you think that is better style, I will change it.

eqvinox replied:
I think the explicit `{-1, 0, 1}` is preferable, because it's easy to fail to consider the consequences of overflow.  While that might not matter in this case, consistent style helps prevent other cases.

Performance-wise - CPU cycles are mostly free.  Memory access is costly.  Capping it to `{-1, 0, 1}` is the former, so I'd refuse performance arguments :smile:.

mjstapp replied:
I'm fine with David's preference - just wanted to ask the question.

mjstapp replied:
so ... I don't think I'm making a "performance argument" - I'm just pointing out that we're doing some special stuff to get {-1, 0, 1} in some paths, but then falling through to `memcmp()`, which does ... arithmetic. it's ok with me, personally, but I wouldn't want someone to be tempted to rely on anything other than {<0, 0, >0} in the result.

mjstapp commented on lib/nexthop.c:
```diff
@@ -36,6 +36,63 @@
+	if (next1->vrf_id > next2->vrf_id)
+		return 1;
+
+	if (next1->type < next2->type)
```
same question as with the vrf_ids - do we need to test twice?

mjstapp replied:
Closing the question - we'll stick with explicit return values

mjstapp commented on lib/nexthop.c:
```diff
@@ -36,6 +36,63 @@
+		break;
+	case NEXTHOP_TYPE_IPV6:
+		ret = memcmp(&next1->gate, &next2->gate, sizeof(union g_addr));
+		if (!ret)
```
shouldn't it be `(ret != 0)` here (and below) ?

sworleys replied:
Yes, I think it should.

mjstapp commented on lib/nexthop.c:
```diff
@@ -36,6 +36,63 @@
+
+	ret = memcmp(&next1->src, &next2->src, sizeof(union g_addr));
+	return ret;
+}
```
we need to test the label stack(s) also?

mjstapp commented on lib/nexthop.c:
```diff
@@ -36,6 +36,63 @@
 DEFINE_MTYPE_STATIC(LIB, NEXTHOP, "Nexthop")
 DEFINE_MTYPE_STATIC(LIB, NH_LABEL, "Nexthop label")
 
+int nexthop_cmp(const struct nexthop *next1, const struct nexthop *next2)
```
Great to have this - but: once it's ready, could we also switch `nexthop_same()` to use it, so we have a single place to maintain the comparison rules?

eqvinox replied:
+1 to this.

(note: the new datastructure APIs also get rid of the need for a separate, "inverted" comparison function for hash tables; hash tables just use the same -1/0/+1 there)

donaldsharp commented:
We will add some doc to say when to use which compare

sworleys commented:
I updated this according to suggestions. Please note the last [commit](https://github.com/FRRouting/frr/pull/4322/commits/8013919c3e12b63d560849d767c9ce779d869000) I removed a function that doesn't really seem to do anything in different from the other `nexthop_same()`.

mjstapp commented on lib/nexthop.c:
```diff
@@ -36,40 +36,94 @@
+
+	ret = memcmp(&next1->src, &next2->src, sizeof(union g_addr));
+	if (ret)
+		return ret;
```
perhaps a small thing, but : I'd prefer to see `ret != 0`

mjstapp commented on lib/nexthop.c:
```diff
@@ -121,27 +175,12 @@ const char *nexthop_type_to_str(enum nexthop_types_t nh_type)
-	if (memcmp(nhl1->label, nhl2->label, nhl1->num_labels))
-		return 0;
+	if (nexthop_labels_cmp(nh1, nh2))
+		return false;
```
again, I'd find it a bit clearer to see `!= 0`

sworleys commented:
> Had a couple of questions about converting from return values to booleans

Force pushed to address these requests.

rwestphal commented on lib/nexthop.c:
```diff
@@ -36,40 +36,94 @@
-			return 0;
-		if (next1->ifindex != next2->ifindex)
-			return 0;
+		ret = memcmp(&next1->gate, &next2->gate, sizeof(union g_addr));
```
I don't like this very much since we're doing a full 16-byte comparison even for IPv4 addresses, and for that we assume that the g_addr union was zero-initialized. But this should be fine since we always create nexthops using `nexthop_new()`, which allocates memory using `XCALLOC`.

mjstapp commented on lib/nexthop.c:
```diff
@@ -36,40 +36,124 @@
+}
+
+int nexthop_cmp(const struct nexthop *next1, const struct nexthop *next2,
+		bool ignore_labels)
```
oh man - you're going to hate this: I dislike adding this to the signature - especially if "false" means "yes, do everything." I think I understand Renato's example - but I think the issue is that we don't properly construct a complete nexthop down in the netlink code in the 'delete' path - we don't even try to look for labels there. in the KERNEL delete path, we might do better to just consider the uniqueness to be type + metric, and skip the nexthop comparison.

If there really are cases where two nexthops that don't have the same label stack are considered "the same", I guess it'd be ok to have a nexthop_cmp_no_labels() or nexthop_same_no_labels() api for that. but it's hard to understand how that could really be correct - aren't those really not the same nexthop?

sworleys replied:
Actually, in reference to @rwestphal 's example.

It looks like in `zebra/rt_netlink.c` we are not getting the labels from the kernel with a route delete. We just ignore them. So maybe that would fix the issue he is talking about it without adding this code edge case.

sworleys replied:
Deleted my comment once I saw yours. Could I not just add code in the kernel delete path to actually parse the labels and add them to it?

sworleys replied:
If I do that, I believe I can just get rid of the edge case bool all together.

rwestphal replied:
> [...] in the KERNEL delete path, we might do better to just consider the uniqueness to be type + metric, and skip the nexthop comparison.

Unfortunately I think this wouldn't be possible. The kernel doesn't use route replace semantics when informing the userspace about the removal of a nexthop of an ECMP route. This means that, upon receipt of a RTM_DELETE message, we need to find the exact nexthop we need to remove from the RIB.

> Could I not just add code in the kernel delete path to actually parse the labels and add them to it?
> Deleted my comment once I saw yours. Could I not just add code in the kernel delete path to actually parse the labels and add them to it?

That would solve the problem, but you'd need to do the same thing in _zebra/kernel_socket.c_. The problem, however, is that we don't know if other kernels (e.g. OpenBSD) send route delete messages containing MPLS labels, and this renders this solution a bit fragile.

Also, being able to do partial nexthop comparisons might be useful for other things in the future. One example that comes to mind is implementing a "nexthop" option to the "show ip route" command. Some vendors implement this option, which allows you to see only the routes that point to a specific nexthop, regardless if it has MPLS labels or not.

mjstapp replied:
I didn't mean to imply that I didn't want to offer a more limited comparison - just that:
a) the specific example had some other aspects to it, and
b) I'd rather offer the special case as a special api, like `nexthop_same_no_labels` so that it's really clear what's happening, and so that every other user of `nexthop_same` doesn't have to figure out how to use an extra boolean arg.

rwestphal replied:
Understood. The only problem I see with `nexthop_same_no_labels` is that it can lead to a bit of code duplication, but this might be acceptable in this case.

Other option would be to rename the `ignore_labels` parameter to `flags`, and add flags like `F_NHCMP_IGN_LABELS`. This would give us more flexibility to perform other kinds of comparisons in the future (e.g. ignore ifindexes?), but admittedly we might not need all this flexibility.

I'll approve this PR now as it looks good to me already (the `rib_delete()` bug was fixed), regardless if you want to do further changes to improve the API.

mjstapp replied:
Thanks, Renato - all of that sounds sensible.
totally agree about duplication: I'd suggest taking the body of the implementation and making it `_internal()` and `static`. then the public apis - `nexthop_same()` and `nexthop_same_no_labels()` - would just pass through to the shared implementation. If in the future we want more flavors, we can support that too, while having a clear public api that's easy to use correctly.

rwestphal replied:
> I'd suggest taking the body of the implementation and making it `_internal()` and `static`. then the public apis - `nexthop_same()` and `nexthop_same_no_labels()` - would just pass through to the shared implementation.

Sounds good to me.

sworleys commented:
Force pushed to create `nexthop_same_no_labels()` rather than passing a bool. Also, changed internal function names to have `_` before them.

I prevented duplication by making a `nexthop_cmp_no_labels()` that `nexthop_cmp()` calls as well. It just then separately checks labels with `_nexthop_labels_cmp()`

sworleys commented:
Rebased to get rid of deprecation warning.
